File: dsa_leetcode_beginner/0269-alien-dictionary/0269-alien-dictionary.py
def alienOrder(words):
    # Step 1: Create the graph.
    adj = {c: set() for word in words for c in word}
    
    for i in range(len(words) - 1):
        word1, word2 = words[i], words[i + 1]
        minLen = min(len(word1), len(word2))
        
        # Handle the case where order is impossible.
        if len(word1) > len(word2) and word1[:minLen] == word2[:minLen]:
            return ""  
        
        for j in range(minLen):
            if word1[j] != word2[j]:
                adj[word1[j]].add(word2[j])
                break
    
    # Step 2: Detect cycles and perform DFS.
    visited = {}  # False = visited but not processed, True = processed.
    order = []
    
    def dfs(c):
        if c in visited:
            return visited[c]  # If True, no cycle; if False, cycle detected.
        visited[c] = False
        for nei in adj[c]:
            if dfs(nei) == False:
                return False  # Cycle detected.
        visited[c] = True
        order.append(c)
        return True
    
    for c in adj:
        if c not in visited:
            if not dfs(c):
                return ""  # Cycle detected.
    
    # Reverse the post-visit ordered list.
    return "".join(order[::-1])  


#Question: https://www.lintcode.com/problem/892/
#Blog: https://blog.unwiredlearning.com/alien-dictionary

File: dsa_leetcode_beginner/0269-alien-dictionary/0269-alien-dictionary.java

import java.util.*;

public class AlienDictionary {
    public String alienOrder(String[] words) {
        // Step 1: Create the graph.
        Map<Character, Set<Character>> adj = new HashMap<>();
        for (String word : words) {
            for (char c : word.toCharArray()) {
                adj.putIfAbsent(c, new HashSet<>());
            }
        }

        for (int i = 0; i < words.length - 1; i++) {
            String word1 = words[i];
            String word2 = words[i + 1];
            int minLen = Math.min(word1.length(), word2.length());

            // Handle the case where order is impossible.
            if (word1.length() > word2.length() && word1.substring(0, minLen).equals(word2.substring(0, minLen))) {
                return "";
            }

            for (int j = 0; j < minLen; j++) {
                if (word1.charAt(j) != word2.charAt(j)) {
                    adj.get(word1.charAt(j)).add(word2.charAt(j));
                    break;
                }
            }
        }

        // Step 2: Detect cycles and perform DFS.
        Map<Character, Boolean> visited = new HashMap<>();
        List<Character> order = new ArrayList<>();

        for (char c : adj.keySet()) {
            if (!visited.containsKey(c)) {
                if (!dfs(c, adj, visited, order)) {
                    return ""; // Cycle detected.
                }
            }
        }

        // Reverse the post-visit ordered list.
        Collections.reverse(order);
        StringBuilder result = new StringBuilder();
        for (char c : order) {
            result.append(c);
        }
        return result.toString();
    }

    private boolean dfs(char c, Map<Character, Set<Character>> adj, Map<Character, Boolean> visited, List<Character> order) {
        if (visited.containsKey(c)) {
            return visited.get(c); // If True, no cycle; if False, cycle detected.
        }
        visited.put(c, false);
        for (char nei : adj.get(c)) {
            if (!dfs(nei, adj, visited, order)) {
                return false; // Cycle detected.
            }
        }
        visited.put(c, true);
        order.add(c);
        return true;
    }
}
